src/invert/laplace/impls/pcr/pcr.cxx

```diff
@@ -0,0 +1,1137 @@
+        int iy = ys + ind % ny;
+
+        // Take FFT in Z direction, apply shift, and put result in k1d
+
+        if (((ix < inbndry) && (inner_boundary_flags & INVERT_SET) && localmesh->firstX()) ||
+            ((localmesh->LocalNx - ix - 1 < outbndry) && (outer_boundary_flags & INVERT_SET) &&
+             localmesh->lastX())) {
+          // Use the values in x0 in the boundary
+          rfft(x0(ix, iy), localmesh->LocalNz, std::begin(k1d));
+        } else {
+          rfft(rhs(ix, iy), localmesh->LocalNz, std::begin(k1d));
+        }
+
+        // Copy into array, transposing so kz is first index
+        for (int kz = 0; kz < nmode; kz++)
+          bcmplx3D((iy - ys) * nmode + kz, ix - xs) = k1d[kz];
+      }
+
+      // Get elements of the tridiagonal matrix
+      // including boundary conditions
+      BOUT_OMP(for nowait)
+      for (int ind = 0; ind < nsys; ind++) {
+        // ind = (iy - ys) * nmode + kz
+        int iy = ys + ind / nmode;
+        int kz = ind % nmode;
+
+        BoutReal kwave = kz * 2.0 * PI / (coords->zlength()); // wave number is 1/[rad]
+        tridagMatrix(&a3D(ind, 0), &b3D(ind, 0), &c3D(ind, 0), &bcmplx3D(ind, 0), iy,
+                     kz,    // True for the component constant (DC) in Z
+                     kwave, // Z wave number
+                     global_flags, inner_boundary_flags, outer_boundary_flags, &Acoef,
+                     &C1coef, &C2coef, &Dcoef,
+                     false); // Don't include guard cells in arrays
+      }
+    }
+
+    // Solve tridiagonal systems
+    //cr->setCoefs(a3D, b3D, c3D);
+    //cr->solve(bcmplx3D, xcmplx3D);
+    cr_pcr_solver(a3D,b3D,c3D,bcmplx3D,xcmplx3D);
+    //verify_solution(a3D,b3D,c3D,bcmplx3D,xcmplx3D);
+
+    // FFT back to real space
+    BOUT_OMP(parallel) {
+      /// Create a local thread-scope working array
+      auto k1d = Array<dcomplex>((localmesh->LocalNz) / 2 +
+                                 1); // ZFFT routine expects input of this length
+
+      const bool zero_DC = global_flags & INVERT_ZERO_DC;
+
+      BOUT_OMP(for nowait)
+      for (int ind = 0; ind < nxny; ++ind) { // Loop over X and Y
+        // ind = (ix - xs)*(ye - ys + 1) + (iy - ys)
+        int ix = xs + ind / ny;
+        int iy = ys + ind % ny;
+
+        if (zero_DC) {
+          k1d[0] = 0.;
+        }
+
+        for (int kz = zero_DC; kz < nmode; kz++)
+          k1d[kz] = xcmplx3D((iy - ys) * nmode + kz, ix - xs);
+
+        for (int kz = nmode; kz < localmesh->LocalNz / 2 + 1; kz++)
+          k1d[kz] = 0.0; // Filtering out all higher harmonics
+
+        irfft(std::begin(k1d), localmesh->LocalNz, x(ix, iy));
+      }
+    }
+  }
+
+  checkData(x);
+
+  return x;
+}
+
+
+/** 
+ * @brief   Initialize local private variables from global input parameters.
+ * @param   n Size of global array
+ * @param   np_world Number of MPI process
+ * @param   rank_world rank ID in MPI_COMM_WORLD
+*/
+void LaplacePCR :: setup(int n, int np_world, int rank_world)
```

Comment:
Can this be inlined into the constructor? It doesn't look like it's called anywhere else